Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+    ///            header fragments, etc.
+    /// - Throws: Errors returned from HPACK encoder.
+    func encode(frame: HTTP2Frame, to buf: inout ByteBuffer) throws -> [IOData] {
+        // note our starting point
+        let start = buf.writerIndex
+        
+//      +-----------------------------------------------+
+//      |                 Length (24)                   |
+//      +---------------+---------------+---------------+
+//      |   Type (8)    |   Flags (8)   |
+//      +-+-------------+---------------+-------------------------------+
+//      |R|                 Stream Identifier (31)                      |
+//      +=+=============================================================+
+//      |                   Frame Payload (0...)                      ...
+//      +---------------------------------------------------------------+
+        
+        // skip 24-bit length for now, we'll fill that in later
+        buf.moveWriterIndex(forwardBy: 3)
+        
+        // 8-bit type
+        buf.write(integer: frame.payload.code)
+        // 8-bit flags -- we don't use padding when encoding in NIO, though
+        buf.write(integer: frame.flags.subtracting(.padded).rawValue)
+        // 32-bit stream identifier -- ensuring the top bit is empty
+        buf.write(integer: frame.streamID.safeNetworkStreamID ?? 0)
+        
+        func writePayloadSize(_ size: Int) {
+            var bytes: (UInt8, UInt8, UInt8)
+            bytes.0 = UInt8((size & 0xff_00_00) >> 16)
+            bytes.1 = UInt8((size & 0x00_ff_00) >>  8)
+            bytes.2 = UInt8( size & 0x00_00_ff)
+            withUnsafeBytes(of: bytes) { ptr in
+                _ = buf.set(bytes: ptr, at: start)
+            }
+        }
+        
+        // frame payload follows, which depends on the frame type itself
+        switch frame.payload {
+        case .data(let data):
+            writePayloadSize(data.readableBytes)
+            return [data]
+            
+        case .headers(let headers, let priority):
+            var payloadSize = 0
+            if let priority = priority {
+                // TODO: throw the proper error here, or determine if we can assume it's valid
+                var dependencyRaw = UInt32(priority.dependency.networkStreamID ?? 0)
+                if priority.exclusive {
+                    dependencyRaw |= 0x8000_0000
+                }
+                buf.write(integer: dependencyRaw)
+                buf.write(integer: priority.weight)
+                payloadSize += 5
+            }
+            
+            // HPACK-encoded data. Find the stream and its associated encoder/decoder
+            let stream = HTTP2StreamData.getStream(for: frame.streamID, allocator: self.allocator)
+            try stream.encoder.beginEncoding(allocator: self.allocator)
+            try stream.encoder.append(headers: headers)
+            let encoded = try stream.encoder.endEncoding()
+            
+            payloadSize += encoded.readableBytes
```

Comment:
You probably don't need to do this math in each of these cases. Instead you should be able to backfill by checking the number of `readableBytes` and subtracting the space you left at the start for the frame header.